gems/sorbet-runtime/test/types/props/optional.rb

```diff
@@ -32,6 +35,24 @@ def self.prop2_source
     assert_equal(52, DefaultsStruct.prop2_source)
   end
 
+  it 'shares structure where appropriate' do
+    DefaultsStruct.prop2_source = 0
+    m1 = DefaultsStruct.new
+    m2 = DefaultsStruct.new
+
+    # Same object
+    assert_same(m1.prop1, m2.prop1)
+    assert_same(m1.trueprop, m2.trueprop)
+    assert_same(m1.falseprop, m2.falseprop)
+    assert_same(m1.default_integer, m2.default_integer)
+
+    # Equal, but different object
+    refute_same(m1.default_array, m2.default_array)
+    refute_same(m1.default_mutable_string, m2.default_mutable_string)
+    assert_equal(m1.default_array, m2.default_array)
+    assert_equal(m1.default_mutable_string, m2.default_mutable_string)
+  end
```

Comment:
This test would be a great thing to split out into a separate PR, so it's easy to see that a) this is the current behavior; and b) this PR doesn't change existing behavior.